platform-sdk/docs/proposals/TSS-Ledger-Id/TSS-Ledger-Id.md

```diff
@@ -0,0 +1,993 @@
+   building a new hashgraph from round 0. It is now in post-genesis mode.
+6. If any pre-genesis nodes connects to the post-genesis node, the post-genesis node provides the pre-genesis node a
+   copy of the tss key-material and ledger id. This is the only communication supported with pre-genesis nodes.
+
+#### Startup for Existing Networks
+
+##### The Release Containing the `tssEncryptionKey`
+
+This release must occur with or after the complete deployment of the TSS-Roster proposal.
+
+This release contains all the state data structures needed to store the public `tssEncryptionKey` in the state and all
+code needed to support storing and retrieving the private `tssEncryptionKey` on disk and managing the `tssEncryptionKey`
+life cycle, including key rotation.
+
+The release supporting the addition of the `tssEncryptionKey` to the roster also provides modified HAPI
+transactions for updating the address book with a node's `tssEncryptionKey`. Until the next software upgrade, only
+the candidate roster will have the `tssEncryptionKey` possibly set in roster entries.
+
+The following startup sequence change is included in this software release:
+
+1. The app reads the state and cryptography from disk (including support for reading the private `tssEncryptionKey`)
+2. The app hands the state and private keys to the platform.
+3. If a software upgrade is detected, if the candidate roster is set, it is adopted immediately. If there is no
+   candidate roster, the existing active roster in the state remains the active roster.
+4. If a new active roster is set and the node's entry in the new active roster contains a public `tssEncryptionKey`,
+   it is validated against the loaded private `tssEncryptionKey`. If there is a mismatch, or the private key was
+   not loaded, a critical error is logged and the node shuts down.
+5. The rest of the startup sequence is consistent with the TSS-Roster proposal.
+
+##### The Release Containing the TSS Bootstrap Process
+
+This release must be at least 1 release after the `tssEncryptionKey` release.
+
+The following startup sequence occurs: (This sequence should work for this release and all subsequent releases
+unless the startup sequence is modified by subsequent releases.)
+
+1. The app reads the state and cryptography from disk (including support for reading the private `tssEncryptionKey`)
+2. The app hands the state and private keys to the platform.
+3. The platform validates that its private `tssEncryptionKey`  matches its public `tssEncryptionKey` in the active
+   roster. If there is a mismatch, a critical error is logged and the node shuts down.
+4. If a software upgrade is detected,
+   1. If the existing active roster and candidate roster do not have key material, adopt the candidate roster
+      immediately.
+   2. If the candidate roster exists, has key material, and the number of yes votes passes the threshold for
+      adoption, then the candidate roster is adopted.
```

Comment:
Something doesn't look right here for as long as we have the first option. Suppose the app installs a candidate roster with an intention to bootstrap TSS by collecting the key material, but the network happens to go through a software upgrade before the collection could start. The first option would adopt the candidate roster immediately while this isn't what the app intended to do.

Reply:
The very first time that the program starts with the new code, the candidate roster will not have any key material.  The prior software version didn't have the code to generate the key material. 

If this new software deployment can't generate any key material before the software upgrade occurs, either we did not do enough testing, or something went really wrong with TSS, but we shouldn't prevent the adoption of the roster.   We should continue the same behavior as the previous software deployment, which is to adopt the roster on software upgrade.

Reply:
I see arguments from both the sides, but I still don't think it is right for a TSS-capable version of the software to adopt a candidate roster that doesn't have TSS material collected as this may contradict the intention of the system owners who installed the candidate roster in the first place before upgrading to that version of the software.

Could we get a clarification from product managers and/or engineering leadership as to what the right solution here is?

Reply:
I've updated with clarification from leadership that until we key the network with a ledger id for the first time, the previous semantics stay the same, to always adopt the candidate roster on software upgrade.

Reply:
Thanks! With the clarifications that we've received and the updated wording, this part now looks fine.